src/components/KYCWall.js

```diff
@@ -0,0 +1,127 @@
+import CONST from '../CONST';
+import Navigation from '../libs/Navigation/Navigation';
+import AddPaymentMethodMenu from './AddPaymentMethodMenu';
+import getClickedElementLocation from '../libs/getClickedElementLocation';
+import * as PaymentUtils from '../libs/PaymentUtils';
+import * as PaymentMethods from '../libs/actions/PaymentMethods';
+import ONYXKEYS from '../ONYXKEYS';
+import userWalletPropTypes from '../pages/EnablePayments/userWalletPropTypes';
+import Log from '../libs/Log';
+
+const propTypes = {
+    /** Route for the Add Bank Account screen for a given navigation stack */
+    addBankAccountRoute: PropTypes.string.isRequired,
+
+    /** Route for the Add Debit Card screen for a given navigation stack */
+    addDebitCardRoute: PropTypes.string.isRequired,
+
+    /** Route for the KYC enable payments screen for a given navigation stack */
+    enablePaymentsRoute: PropTypes.string.isRequired,
+
+    ...userWalletPropTypes,
+};
+
+const defaultProps = {
+    // eslint-disable-next-line react/default-props-match-prop-types
+    userWallet: {},
+};
+
+// This component allows us to block various actions by forcing the user to first add a default payment method and successfully make it through our Know Your Customer flow
+// before continuing to take whatever action they originall intended to take. It requires a button as a child and a native event so we can get the coordinates and use it
+// to render the AddPaymentMethodMenu in the correct location.
+class KYCWall extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.continue = this.continue.bind(this);
+
+        this.state = {
+            shouldShowAddPaymentMenu: false,
+            anchorPositionTop: 0,
+            anchorPositionLeft: 0,
+        };
+    }
+
+    componentDidMount() {
+        PaymentMethods.kycWallRef.current = this;
+    }
+
+    componentWillUnmount() {
+        PaymentMethods.kycWallRef.current = null;
+    }
+
+    /**
+     * Take the position of the button that calls this method and show the Add Payment method menu when the user has no valid payment method.
+     * If they do have a valid payment method they are navigated to the "enable payments" route to complete KYC checks.
+     * If they are already KYC'd we will continue whatever action is gated behind the KYC wall.
+     *
+     * @param {Event} event
+     */
+    continue(event) {
+        // Check to see if user has a valid payment method on file and display the add payment popover if they don't
+        if (!PaymentUtils.hasExpensifyPaymentMethod(this.props.cardList, this.props.bankAccountList)) {
+            Log.info('[KYC Wallet] User does not have valid payment method');
+            const position = getClickedElementLocation(event.nativeEvent);
+            this.setState({
+                shouldShowAddPaymentMenu: true,
+                anchorPositionTop: position.bottom - 226,
+                anchorPositionLeft: position.right - 356,
```

Comment:
why the hardcode? should these be CONSTs?

Reply:
Ah hmm, I guess they are hardcoded because I couldn't think of a better way to position this and forgot to go back and improve this so thanks for catching it!

I think maybe the more appropriate thing would be to set the bottom of the popover to some distance away from the top of the button and then set a constant for the overall height of the button. Feels like we did something similar here for the emoji picker.

https://github.com/Expensify/App/blob/021ad5d5b057fcecd98ee70b6bf52938cca5fe32/src/pages/home/report/ReportActionCompose.js#L665-L668

Reply:
Oh hmm it only will give us bottom and right actually:

<img width="1009" alt="2022-01-06_13-14-21" src="https://user-images.githubusercontent.com/32969087/148465672-77e3c977-1aee-4065-89b0-eea60c21147d.png">
 
The implementation is pretty complicated for how we calculate these for the emojis eep actually since we

1. Measure the vertical/horizontal placement of the button
2. Re-measure when the window changes
3.  Use that + the size of the picker to display it correctly

Gonna take me a second to create that same behavior here.

Reply:
Actually might do it in a follow up. It's gonna be too much to do right now.

Reply:
Fixed these up with CONST for now.

Reply:
Gonna tackle here https://github.com/Expensify/Expensify/issues/191203